### PR TITLE
feature(files): adds a service for serving files from filestore

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds API and handlers for serving files from the filestore
- [x] How to handle file entities with custom filestore?
- [x] Write docs
- [x] Write tests
- [ ] Address comment https://github.com/Elgg/Elgg/pull/9147#discussion_r48444138

### CODE REVIEW

**Comment by hypeJunction** [comment]
This looks quite nice, I think. Let me know what you think, and we can work on cleaning this up, adding tests and docs. We should also avoid booting engine fully for this. Perhaps we should add an option in `Elgg\Application::bootCore($boot_plugins = true)`?

File: engine/classes/Elgg/FileService/File.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+	/**
+	 * Sets content disposition
+	 *
+	 * @param string $disposition Content disposition ('inline' or 'attachment')
+	 * @return void
+	 */
+	public function setDisposition($disposition = self::DISPOSITION_ATTACHMENT) {
+		if (!in_array($disposition, array(self::DISPOSITION_ATTACHMENT, self::DISPOSITION_INLINE))) {
+			throw new \UnexpectedValueException("Disposition $disposition is not supported in " . __CLASS__);
```
Comment (mrclay) [inline engine/classes/Elgg/FileService/File.php]: InvalidArgumentException

File: engine/classes/Elgg/FileService/File.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+		$data = array(
+			'expires' => isset($this->expires) ? $this->expires : time() + self::DEFAULT_TTL,
+			'last_updated' => filemtime($this->file->getFilenameOnFilestore()),
+			'disposition' => $this->disposition == self::DISPOSITION_INLINE ? 'i' : 'a',
+			'path' => $relative_path,
+		);
+
+
+		$key = _elgg_services()->siteSecret->get();
```
Comment (mrclay) [inline engine/classes/Elgg/FileService/File.php]: Use `get(true)` (binary key preferred over ASCII encoding). Or just use `elgg_build_hmac($data)->getToken()` which does this.

**Comment by mrclay** [comment]
Time to replace some core URLs with this API? I don't think we could merge without some real testing.

**Comment by hypeJunction** [comment]
Yes, will make some time tomorrow. I want this in badly.

File: engine/lib/filestore.php [hunk 3]
```diff
@@ -581,6 +581,43 @@ function _elgg_filestore_test($hook, $type, $value) {
...
+/**
+ * Returns file's URL for inline display
+ * Suitable for displaying cacheable resources, such as user avatars
+ *
+ * @param \ElggFile $file       File object or entity
+ * @param bool      $use_cookie Limit URL validity to current session only
+ * @param string    $expires    URL expiration, as a string suitable for strtotime()
+ * @return string
+ */
+function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = '+1 year') {
```
Comment (juho-jaakkola) [inline engine/lib/filestore.php]: I wonder which is better API:

```php
elgg_get_inline_url($file, false, '+1 year');
```
or
```php
elgg_get_url(array(
    'file' => $file,
    'use_cookie' => false,
    'expires' => '+1 year',
    'disposition' => 'inline',
));
```

File: engine/classes/Elgg/Application/ServeFileHandler.php [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	/**
+	 * Send a 304 and exit() if the ETag matches the request
+	 *
+	 * @param string $etag ETag value
+	 * @return void
+	 */
+	protected function handle304($etag) {
+		// If is the same ETag, content didn't change.
```
Comment (juho-jaakkola) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: ```If ETag is the same, content hasn't changed.```

This old comment has bugged me more than it should. :)

File: engine/lib/filestore.php (hunk 3 above)
Comment (hypeJunction) [inline engine/lib/filestore.php]: I think the idea is to not make devs think about it too much and let core handle the internals.

**Comment by hypeJunction** [comment]
File icons and downloads, as well as user avatars are handled with the new API.
I have also made the following change:
 * we need to validate that mtime of the file matches the one in the URL
 * we need to set default expire to 0 for inline URLs so that the URL structure is the same on consecutive requests and we can benefit from cache
 * fixed bugs with etag, so we now get proper 304 headers

File: engine/classes/Elgg/Application/ServeFileHandler.php [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		}
+
+		$etag = '"' . $last_updated . '"';
+		$this->handle304($etag);
+
+		// @todo: change to minimal boot without plugins
+		$this->application->bootCore();
+
+
+		$key = _elgg_services()->siteSecret->get();
```
Comment (hypeJunction) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Remove, no longer needed.

File: engine/classes/Elgg/Application/ServeFileHandler.php [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+			header("Content-disposition: inline");
+		} else {
+			$basename = basename($filenameonfilestore);
+			header("Content-disposition: attachment; filename='$basename'");
+		}
+
+		if ($expires) {
+			$expires_str = gmdate('D, d M Y H:i:s \G\M\T', $expires);
+		} else {
+			$expires_str = gmdate('D, d M Y H:i:s \G\M\T', strtotime("+3 years"));
```
Comment (hypeJunction) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Should we use a different default?

File: engine/classes/Elgg/Application/ServeFileHandler.php [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+		}
+
+		header('Expires: ' . $expires_str, true);
+
+		$cache_control = $use_cookie ? 'no-cache' : 'public';
```
Comment (hypeJunction) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Do we need to add max-age?

File: engine/classes/Elgg/Application/ServeFileHandler.php [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+		header("Pragma: $cache_control", true);
+		header("Cache-Control: $cache_control", true);
+
+		header("ETag: $etag");
+
+		readfile($filenameonfilestore);
```
Comment (hypeJunction) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Should we add `X-Sendfile` if Apache and `mod_xsendfile` is enabled?

File: engine/classes/Elgg/FileService/File.php [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * File service
+ * 
+ * @access private
+ */
+class File {
+
+	const DISPOSITION_INLINE = 'inline';
+	const DISPOSITION_ATTACHMENT = 'attachment';
+	const DEFAULT_TTL = 7200;
```
Comment (hypeJunction) [inline engine/classes/Elgg/FileService/File.php]: No longer using this

File: engine/classes/Elgg/FileService/File.php [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+namespace Elgg\FileService;
+
+/**
+ * File service
+ * 
+ * @access private
+ */
+class File {
+
+	const DISPOSITION_INLINE = 'inline';
```
Comment (hypeJunction) [inline engine/classes/Elgg/FileService/File.php]: Should we shorten to `INLINE` and `ATTACHMENT`?

File: mod/file/start.php [hunk 11]
```diff
@@ -359,66 +359,86 @@ function file_set_url($hook, $type, $url, $params) {
...
-		} else {
-			$mime = 'none';
-			$base_type = 'none';
+		if (!empty($thumbfile)) {
+			$readfile = new ElggFile();
+			$readfile->owner_guid = $file->owner_guid;
+			$readfile->setFilename($thumbfile);
+			$thumb_url = elgg_get_inline_url($readfile, true);
+			if ($thumb_url) {
+				return $thumb_url;
```
Comment (hypeJunction) [inline mod/file/start.php]: We don't have an icon for simpletype = image. Need to fallback to default entity icon?

File: mod/file/start.php [hunk 12]
```diff
@@ -359,66 +359,86 @@ function file_set_url($hook, $type, $url, $params) {
...
+			case "medium":
+				$thumbfile = $file->smallthumb;
+				break;
+			case "large":
+			default:
+				$thumbfile = $file->largethumb;
+				break;
 		}
 
-		$mapping = array(
```
Comment (hypeJunction) [inline mod/file/start.php]: I terminated early, and this is just indentation here.

File: mod/file/thumbnail.php [hunk 13]
```diff
@@ -1,10 +1,10 @@
 <?php
+
```
Comment (hypeJunction) [inline mod/file/thumbnail.php]: I say we deprecate this file and get rid of it in 3.x

File: mod/file/views/default/resources/file/download.php [hunk 14]
```diff
@@ -15,23 +15,4 @@
...
-header("Content-Disposition: attachment; filename=\"$filename\"");
-header("Content-Length: {$file->getSize()}");
-
-while (ob_get_level()) {
-    ob_end_clean();
-}
-flush();
-readfile($file->getFilenameOnFilestore());
-exit;
+forward(elgg_get_download_url($file));
```
Comment (hypeJunction) [inline mod/file/views/default/resources/file/download.php]: Deprecate this view?

File: views/default/resources/avatar/view.php [hunk 15]
```diff
@@ -12,39 +12,13 @@
...
-	elgg_log("Unable to get avatar for user with GUID $user_guid", 'ERROR');
-}
-
-if (!$success) {
-	forward(elgg_get_simplecache_url("icons/user/default{$size}.gif"));
+if ($user) {
+	$filehandler = new ElggFile();
+	$filehandler->owner_guid = $user->guid;
+	$filehandler->setFilename("profile/{$user->guid}{$size}.jpg");
+	$avatar_url = elgg_get_inline_url($filehandler);
```
Comment (hypeJunction) [inline views/default/resources/avatar/view.php]: This can return `false`

File: engine/classes/Elgg/Application/ServeFileHandler.php [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+		if (!is_readable($filenameonfilestore)) {
+			$this->send404();
+		}
+
+		$actual_last_updated = filemtime($filenameonfilestore);
+		if ($actual_last_updated != $last_updated) {
+			$this->send403('URL has expired');
+		}
+
+		$mime = $this->getContentType($path_from_dataroot);
```
Comment (hypeJunction) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: this should be `$filenameonfilestore`

File: engine/classes/Elgg/Application/ServeFileHandler.php (hunk 8 above)
Comment (mrclay) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Let's tackle that later. Already a ticket for it I bet.

File: engine/classes/Elgg/Application/ServeFileHandler.php (hunk 7 above)
Comment (mrclay) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Expires should be sufficient, but for HTTP/1.1, yeah let's add it (only on Cache-Control, not Pragma)

File: engine/classes/Elgg/FileService/File.php (hunk 10 above)
Comment (mrclay) [inline engine/classes/Elgg/FileService/File.php]: Sure.

File: engine/classes/Elgg/Application/ServeFileHandler.php (hunk 6 above)
Comment (mrclay) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: I always see 1 year. Doesn't matter at all :)

File: mod/file/start.php (hunk 11 above)
Comment (mrclay) [inline mod/file/start.php]: What's the current code do in this case?

Comment (mrclay) [inline mod/file/start.php]: We can just file a bug and deal later.

File: mod/file/thumbnail.php (hunk 13 above)
Comment (mrclay) [inline mod/file/thumbnail.php]: Sure.

File: mod/file/views/default/resources/file/download.php (hunk 14 above)
Comment (mrclay) [inline mod/file/views/default/resources/file/download.php]: :+1:

File: mod/profile/start.php [hunk 17]
```diff
@@ -107,34 +103,7 @@ function profile_set_url($hook, $type, $url, $params) {
...
-
-	try {
-		if ($filehandler->exists()) {
-			return "mod/profile/icondirect.php?lastcache=$icon_time&guid=$user_guid&size=$size";
-		}
-	} catch (InvalidParameterException $e) {
-		elgg_log("Unable to get profile icon for user with GUID $user_guid", 'ERROR');
-		return elgg_get_simplecache_url("icons/default/$size.png");
-	}
+	elgg_deprecated_notice("Profile plugin no longer customizes avatar url using 'entity:icon:url' hook", '2.1');
```
Comment (mrclay) [inline mod/profile/start.php]: Good!

File: mod/file/start.php (hunk 11 above)
Comment (hypeJunction) [inline mod/file/start.php]: It always return a URL, which then silently fails.

File: engine/classes/Elgg/Application/ServeFileHandler.php (hunk 7 above)
Comment (hypeJunction) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Hm, I don't think I understand cache control enough. I will let it sit for now, and we can add it later.

**Comment by hypeJunction** [comment]
Re: files with custom filestore. I don't think file plugin is capable of handling such cases anyway, and anyone implementing custom filestore would have to rewrite all handlers and use hooks to implement custom serving logic.
Re: tests. What is the strategy? I don't think we can actually test the ServeFileHandler. FileService\File would probably just testing a bunch of combinations and asserting that the actual URL matches and expected URL?

File: engine/classes/Elgg/FileService/File.php [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+			'serve-file',
+			"e{$data['expires']}",
+			"l{$data['last_updated']}",
+			"d{$data['disposition']}",
+			"c{$data['use_cookie']}",
+			$mac,
+			$relative_path,
+		);
+
+		return elgg_normalize_url(implode('/', $url_segments));
```
Comment (hypeJunction) [inline engine/classes/Elgg/FileService/File.php]: The line was too long erroring Travis tests.

File: engine/classes/Elgg/Application/ServeFileHandler.php [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+	/**
+	 * Send a 304 and exit() if the ETag matches the request
+	 *
+	 * @param string $etag ETag value
+	 * @return void
+	 */
+	protected function handle304($etag) {
+		if (isset($this->server_vars['HTTP_IF_NONE_MATCH'])
+			&& trim($this->server_vars['HTTP_IF_NONE_MATCH']) === $etag) {
+			header("HTTP/1.1 304 Not Modified");
```
Comment (juho-jaakkola) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: I guess tests could use ```headers_list()``` for testing these ```handleNNN()``` methods.

```$this->assertContains("HTTP/1.1 304 Not Modified", headers_list());```

Comment (juho-jaakkola) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Or we could inject ```Symfony\Component\HttpFoundation\Response``` and mock that in the tests.

Comment (hypeJunction) [inline engine/classes/Elgg/Application/ServeFileHandler.php]: Yes, I think we should inject and mock `Symfony\Component\HttpFoundation\Response`. We can then eliminate `exit()` calls.

**Comment by hypeJunction** [comment]
@mrclay @juho-jaakkola I have rewritten the server to use Symfony's `Response` and `BinaryFileResponse`, have added some tests and docs. I am not sure it makes sense to test the `File` separately, as most of it is tested in `ServeFileHandler`.

**Comment by hypeJunction** [comment]
Test failures seem unrelated. Would be good if someone could check.

File: mod/file/actions/file/upload.php [hunk 20]
```diff
@@ -68,6 +68,10 @@
 		// user blanked title, but we need one
 		$title = $file->title;
 	}
+
+	if ($entity->access_id !== $access_id) {
```
Comment (mrclay) [inline mod/file/actions/file/upload.php]: The changes in this file look like a bugfix that could be moved to a separate PR for 2.0, or am I missing something? Also you'd want to set `$reset_icon_urls` either way so you don't get notices below.

File: engine/tests/phpunit/Elgg/Application/ServeFileHandlerTest.php [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+
+		$request = $this->createRequest($file);
+		$response = $this->handler->getResponse($request);
+		$this->assertEquals(403, $response->getStatusCode());
+	}
+
+	/**
+	 * @group FileService
+	 */
+	public function testResponseWhenFileModifiedTime() {
```
Comment (juho-jaakkola) [inline engine/tests/phpunit/Elgg/Application/ServeFileHandlerTest.php]: Test names should describe the expected result. E.g. `testSends403onFileModificationTimeMismatch` or similar.

Comment (juho-jaakkola) [inline engine/tests/phpunit/Elgg/Application/ServeFileHandlerTest.php]: Same for the other tests below.

File: mod/file/actions/file/upload.php (hunk 20 above)
Comment (juho-jaakkola) [inline mod/file/actions/file/upload.php]: I agree.

**Comment by mrclay** [comment]
Strawman: Make a separate download handler like file/download for files with non-default filestore (we'd need a method to determine this). They'd use separate URL that contains GUID instead of path.

File: mod/file/actions/file/upload.php (hunk 20 above)
Comment (hypeJunction) [inline mod/file/actions/file/upload.php]: These changes would make no sense in 2.0. They only make sense for icons handled by the new API. What happens here is we update icons' modified times, so all requests to previously generated URLs will fail due to HMAC mismatch.

**Comment by hypeJunction** [comment]
I think we should get the API in first, and update existing server logic in separate PRs. I will move the 2 commits into separate PRs and add something in developer tools for testing.

**Comment by hypeJunction** [comment]
I have stripped out the changes to avatars and file thumbs. Those are in separate PRs now.
@juho-jaakkola Updated test names.

**Comment by hypeJunction** [comment]
Added `DownloadFileHandler` for files that have custom filestore. It starts the application, retrieves the file and serves it as `attachment`

File: engine/classes/Elgg/FileService/File.php [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+		$root_prefix = _elgg_services()->config->get('dataroot');
+		$path = $this->file->getFilenameOnFilestore();
+		if (substr($path, 0, strlen($root_prefix)) == $root_prefix) {
+			$relative_path = substr($path, strlen($root_prefix));
+		}
+
+		if (!$relative_path) {
+			// File object has a custom filestore
+			if ($this->file->guid) {
+				$url_segments = array(
```
Comment (mrclay) [inline engine/classes/Elgg/FileService/File.php]: This API will let you request a public URL with expires and inline, but depending on the file object, it just won't obey you; you'll get a download link that requires login. We should give URLs that elicit the same behavior. I'll try to send a PR to do that in the next few days.

Comment (hypeJunction) [inline engine/classes/Elgg/FileService/File.php]: Ok. I can use a break from this feature :baby_bottle:

Comment (juho-jaakkola) [inline engine/classes/Elgg/FileService/File.php]: The problem is related only to custom filestores? One option could be to keep using the old URLs for bundled plugins, deprecate support for custom filestores during 2.x, and remove completely in Elgg 3.

Comment (juho-jaakkola) [inline engine/classes/Elgg/FileService/File.php]: @mrclay Can you elaborate the problem you've found?

Comment (mrclay) [inline engine/classes/Elgg/FileService/File.php]: Use the API to generate a public URL that's served inline. What you instead get is a download link that requires login.

Your suggestion is simplest: Kill custom filestores in 3.0 and add this then.

Comment (juho-jaakkola) [inline engine/classes/Elgg/FileService/File.php]: I'm in fact suggesting that we add the API already in 2.x, but we wait until Elgg 3 to use it in bundled plugins, so we don't break any sites using custom filestore.

**Comment by mrclay** [comment]
This will also resolve #5891.

**Comment by juho-jaakkola** [comment]
Anyone against including this in Elgg 2.1 and adding a warning that it won't work with custom filestores? 

Or did @mrclay already have some idea how to make this work with them?

**Comment by mrclay** [comment]
Yeah, let's pull this in for 2.1. See also #9352.

**Comment by mrclay** [comment]
Replaced by #9354
